### PR TITLE
Code fix missing method

### LINKED ISSUE #14329: Add Missing Property QuickFix on Method Call Should Create Method
**TypeScript Version:**  2.2.1 / nightly (2.2.0-dev.201xxxxx)

**Code**

```ts
class Foo {
    constructor() {
        this.bar('abc', 123)
    }
}
```

Get a quick fixes for the call to `bar`. Trigger the quick fix to add a missing property to fix the error

screen shot 2017-02-26 at 11 26 52 pm

**Expected behavior:**
Fix introduces a method that fixes the error

```ts
class Foo {
     bar(p1: string, p2: number) {
          throw new Error('Method not implemented.');
     }

    constructor() {
        this.bar('abc', 123)
    }
}
```

**Actual behavior:**
Fix introduces a `any` property that fixes the error but is not very helpful:

```ts
class Foo {
    bar: any
    constructor() {
        this.bar('abc', 123)
    }
}
```

### PR DESCRIPTION
**Pull request by aozgaa**
Fixes #14329.

Changed the quickfix so it offers to implement the missing member as a method if possible, and offers implementations as property or index signature otherwise.

Should we offer options to implement as a method and a property (ie: one with a function type)?

### CODE REVIEW

**Comment by aozgaa** [comment]
Can you take a look, @vladima, @zhengbli, @mhegazy, @billti ?

File: src/services/codefixes/helpers.ts [hunk 1]
```diff
@@ -86,6 +86,43 @@ namespace ts.codefix {
         }
     }
 
+    export function createBodySignatureFromCallExpression(callExpression: CallExpression, name: string, enclosingDeclaration: ClassLikeDeclaration, checker: TypeChecker): Signature {
+        const newSignatureDeclaration = createNode(SyntaxKind.CallSignature) as SignatureDeclaration;
+        newSignatureDeclaration.parent = enclosingDeclaration;
+        newSignatureDeclaration.name = createNode(SyntaxKind.Identifier) as Identifier;
```
Comment (mhegazy) [inline src/services/codefixes/helpers.ts]: `createIdentifier(name);`

File: src/services/codefixes/helpers.ts [hunk 2]
```diff
@@ -107,36 +144,37 @@ namespace ts.codefix {
...
-            const newParameter = createNode(SyntaxKind.Parameter) as ParameterDeclaration;
+    const optionalToken = createToken(SyntaxKind.QuestionToken);
 
-            newParameter.symbol = new SymbolConstructor(SymbolFlags.FunctionScopedVariable, maxArgsParameterSymbolNames[index] || "rest");
-            newParameter.symbol.valueDeclaration = newParameter;
-            newParameter.symbol.declarations = [newParameter];
-            newParameter.parent = enclosingSignatureDeclaration;
-            if (index >= minArgCount) {
+    function createParameterDeclarationWithoutType(name: string, enclosingSignatureDeclaration: SignatureDeclaration, optional?: boolean): ParameterDeclaration {
+        const newParameter = createNode(SyntaxKind.Parameter) as ParameterDeclaration;
```
Comment (mhegazy) [inline src/services/codefixes/helpers.ts]: `createParameter`

File: src/services/codefixes/helpers.ts [hunk 3]
```diff
@@ -86,6 +86,43 @@ namespace ts.codefix {
...
+            newTypeParameter.parent = newSignatureDeclaration;
+
+            const name = typeArgCount < 8 ? String.fromCharCode(CharacterCodes.T + i) : `T${i}`;
+
+            newTypeParameter.symbol = new SymbolConstructor(SymbolFlags.FunctionScopedVariable, name);
+            newTypeParameter.symbol.valueDeclaration = newTypeParameter;
+            newTypeParameter.symbol.declarations = [newTypeParameter];
+            newTypeParameter.symbol.flags = SymbolFlags.TypeParameter;
+
+            newTypeParameter.name = createNode(SyntaxKind.Identifier) as Identifier;
```
Comment (mhegazy) [inline src/services/codefixes/helpers.ts]: createIdentifer(name)

File: src/services/codefixes/helpers.ts [hunk 4]
```diff
@@ -86,6 +86,43 @@ namespace ts.codefix {
...
+
+        for(let i = 0; i < argCount; i++) {
+            const newParameter = createParameterDeclarationWithoutType(`arg${i}`, newSignatureDeclaration, /*optional*/ false);
+            newSignatureDeclaration.parameters.push(newParameter);
+        }
+        
+        const typeArgCount = callExpression.typeArguments ? callExpression.typeArguments.length : 0;
+        newSignatureDeclaration.typeParameters = createNodeArray<TypeParameterDeclaration>();
+        for(let i = 0; i < typeArgCount; i++) {
+            const newTypeParameter = createNode(SyntaxKind.TypeParameter) as TypeParameterDeclaration;
```
Comment (mhegazy) [inline src/services/codefixes/helpers.ts]: I would add a new function to the factory.ts to createTypeParameter

File: src/services/codefixes/helpers.ts [hunk 5]
```diff
@@ -86,6 +86,43 @@ namespace ts.codefix {
...
+            newTypeParameter.parent = newSignatureDeclaration;
+
+            const name = typeArgCount < 8 ? String.fromCharCode(CharacterCodes.T + i) : `T${i}`;
+
+            newTypeParameter.symbol = new SymbolConstructor(SymbolFlags.FunctionScopedVariable, name);
```
Comment (mhegazy) [inline src/services/codefixes/helpers.ts]: extract these into `createSymbol(flags, name, declarations)`

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -2750,7 +2750,15 @@ namespace ts {
             }
 
             function buildTypeParameterDisplay(tp: TypeParameter, writer: SymbolWriter, enclosingDeclaration?: Node, flags?: TypeFormatFlags, symbolStack?: Symbol[]) {
-                appendSymbolNameOnly(tp.symbol, writer);
+                const declaration = tp.symbol && tp.symbol.declarations && tp.symbol.declarations[0] as TypeParameterDeclaration;
```
Comment (mhegazy) [inline src/compiler/checker.ts]: why do you need these? I would just not create a declaration, and use the `symbol.name`, same for parameters as well..

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -2750,7 +2750,15 @@ namespace ts {
             }
 
             function buildTypeParameterDisplay(tp: TypeParameter, writer: SymbolWriter, enclosingDeclaration?: Node, flags?: TypeFormatFlags, symbolStack?: Symbol[]) {
-                appendSymbolNameOnly(tp.symbol, writer);
+                const declaration = tp.symbol && tp.symbol.declarations && tp.symbol.declarations[0] as TypeParameterDeclaration;
+                if(!declaration) {
+                    appendSymbolNameOnly(tp.symbol, writer);
+                }
+                else {
+                    const text = declaration.name.text;
```
Comment (zhengbli) [inline src/compiler/checker.ts]: Is `declaration.name` guaranteed to exist?

File: src/harness/fourslash.ts [hunk 8]
```diff
@@ -2115,23 +2115,23 @@ namespace FourSlash {
...
-         *
-         * Because codefixes are only applied on the working file, it is unsafe
-         * to apply this more than once (consider a refactoring across files).
+         * Finds and applies a code action corresponding to the supplied parameters.
+         * If index is undefined, applies the unique code action available.
+         * @param errorCode The error code that generated the code action.
+         * @param index The nth (0-index-based) codeaction available generated by errorCode.
          */
-        public verifyRangeAfterCodeFix(expectedText: string, includeWhiteSpace?: boolean, errorCode?: number, index?: number) {
+        public applyCodeAction(errorCode?: number, index?: number) {
```
Comment (zhengbli) [inline src/harness/fourslash.ts]: `getAndApplyCodeAction`

**Comment by aozgaa** [comment]
Putting this on hold while we get `createTypeNode` functionality in.

**Comment by mhegazy** [comment]
closing in favor of https://github.com/Microsoft/TypeScript/pull/16045

**Pull request closed without merging**